### PR TITLE
#206 violation types api

### PR DESCRIPTION
**Pull request by mrandi**
DB:
* join app lifecycle by instance id (add column instance id)
* remove message and add metainfo in violationObject
* add plugin column (full qualified class name)
* add type table with column:
 * id (missing_approvals)
 * help_text (problem and solution as text, will be concatenated with metainfo)
 * severity (high/low as int >= 0)
 * is_audit_relevant (to filter for zmon2)
* new unique constraint ( event_id, account_id, region, type)
* create missing types on startup

API:
* sorting for severity, is_auditing_relevant, type
* <s>grouping for severity, type</s>
* filtering for severity, is_auditing_relevant, type

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="http://app.review.ninja/zalando-stups/fullstop/pull/218" target="_blank">ReviewNinja

File: fullstop-testing/it/fullstop/src/main/resources/schema.sql [hunk 1]
```diff
@@ -67,4 +67,8 @@ CREATE TABLE IF NOT EXISTS fullstop_data.lifecycle (
   version             BIGINT    NOT NULL,
   FOREIGN KEY (application) REFERENCES fullstop_data.application (id),
   FOREIGN KEY (application_version) REFERENCES fullstop_data.app_version (id)
-);
\ No newline at end of file
+);
+
+CREATE TABLE IF NOT EXISTS fullstop_data.violation_type(
+
```
Comment (mrandi) [inline fullstop-testing/it/fullstop/src/main/resources/schema.sql]: add also the property

**Comment by mrandi** [comment]
migrate plugin in order to use also the violation type

File: fullstop-web/fullstop-timemachine/src/test/java/org/zalando/stups/fullstop/swagger/api/FullstopApiTest.java [hunk 2]
```diff
@@ -79,7 +76,7 @@
 
     public static final String ACCOUNT_ID = "123";
 
-    public static final String MESSAGE = "my message";
+    public static final String MESSAGE = "my comment";
```
Comment (mrandi) [inline fullstop-web/fullstop-timemachine/src/test/java/org/zalando/stups/fullstop/swagger/api/FullstopApiTest.java]: change variable name to comment

File: fullstop-testing/it/fullstop/src/main/resources/schema.sql [hunk 3]
```diff
@@ -67,4 +67,8 @@ CREATE TABLE IF NOT EXISTS fullstop_data.lifecycle (
   version             BIGINT    NOT NULL,
   FOREIGN KEY (application) REFERENCES fullstop_data.application (id),
   FOREIGN KEY (application_version) REFERENCES fullstop_data.app_version (id)
+);
+
+CREATE TABLE IF NOT EXISTS fullstop_data.violation_type (
+
```
Comment (mrandi) [inline fullstop-testing/it/fullstop/src/main/resources/schema.sql]: miss something

**Comment by mrandi** [comment]
- [ ] provide a DB migration script
- [ ] tests are missing

File: sql/migration.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  version            BIGINT NOT NULL
+);
+
+
+ALTER TABLE fullstop_data.violation DROP CONSTRAINT IF EXISTS unique_violation;
+
+ALTER TABLE fullstop_data.violation
+ADD COLUMN plugin_fully_qualified_class_name TEXT,
+ADD COLUMN violation_type_entity_id TEXT, -- NOT NULL,
+DROP COLUMN message, -- should we wait for that?
```
Comment (mrandi) [inline sql/migration.sql]: should we wait for the next iteration?

File: sql/migration.sql [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  last_modified_by   TEXT,
+  version            BIGINT NOT NULL
+);
+
+
+ALTER TABLE fullstop_data.violation DROP CONSTRAINT IF EXISTS unique_violation;
+
+ALTER TABLE fullstop_data.violation
+ADD COLUMN plugin_fully_qualified_class_name TEXT,
+ADD COLUMN violation_type_entity_id TEXT, -- NOT NULL,
```
Comment (mrandi) [inline sql/migration.sql]: should we use a default type for old violations?

File: sql/migration.sql (hunk 4 above)
Comment (mrandi) [inline sql/migration.sql]: we will start from scratch, so no migration necessary!

File: sql/migration.sql (hunk 5 above)
Comment (mrandi) [inline sql/migration.sql]: we will start from scratch, so no migration necessary!

**Pull request merged**